extractor/filesystem/language/javascript/bunlock/bunlock.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bunlock extracts bun.lock files
+package bunlock
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/osv-scalibr/extractor"
+	"github.com/google/osv-scalibr/extractor/filesystem"
+	"github.com/google/osv-scalibr/extractor/filesystem/osv"
+	"github.com/google/osv-scalibr/plugin"
+	"github.com/google/osv-scalibr/purl"
+	"golang.org/x/exp/maps"
+
+	"github.com/tailscale/hujson"
+)
+
+type bunLockfile struct {
+	Version  int              `json:"lockfileVersion"`
+	Packages map[string][]any `json:"packages"`
+}
+
+// Extractor extracts npm packages from bun.lock files.
+type Extractor struct{}
+
+// Name of the extractor.
+func (e Extractor) Name() string { return "javascript/bunlock" }
+
+// Version of the extractor.
+func (e Extractor) Version() int { return 0 }
+
+// Requirements of the extractor.
+func (e Extractor) Requirements() *plugin.Capabilities {
+	return &plugin.Capabilities{}
+}
+
+// FileRequired returns true if the specified file matches bun lockfile patterns.
+func (e Extractor) FileRequired(api filesystem.FileAPI) bool {
+	return filepath.Base(api.Path()) == "bun.lock"
+}
+
+// structurePackageDetails returns the name, version, and commit of a package
+// specified as a tuple in a bun.lock
+func structurePackageDetails(a []any) (string, string, string) {
+	str, ok := a[0].(string)
+
+	// todo: should we return an error instead?
+	if !ok {
+		return "", "", ""
+	}
+
+	str, isScoped := strings.CutPrefix(str, "@")
+	name, version, _ := strings.Cut(str, "@")
+
+	if isScoped {
+		name = "@" + name
+	}
+
+	version, commit, _ := strings.Cut(version, "#")
+
+	// bun.lock does not track both the commit and version,
+	// so if we have a commit then we don't have a version
+	if commit != "" {
+		version = ""
+	}
+
+	if strings.HasPrefix(version, "file:") {
+		version = ""
+	}
+
+	return name, version, commit
+}
+
+// Extract extracts packages from bun.lock files passed through the scan input.
+func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
+	var parsedLockfile *bunLockfile
+
+	b, err := io.ReadAll(input.Reader)
+
+	if err != nil {
+		return []*extractor.Inventory{}, fmt.Errorf("could not extract from %q: %w", input.Path, err)
```

Comment:
return nil instead of empty arrays during errors

Reply:
other extractors are returning an empty array rather than nil, which is why I did that here:

![image](https://github.com/user-attachments/assets/c055706d-10ed-464d-9dd8-148f09f86ff9)

Given that, do you want to keep this as-is or should I do a PR updating the other extractors to return `nil` as well?

Reply:
In most places we return nil in the extractors - I think there are some leftover places where we didn't add that so I'd say let's return nil here too.
If you can create a followup PR to modify the other extractors to also return nil that's be great.

Reply:
Have opened #391 for this